### PR TITLE
[CIS-638] Decrease channel launch times

### PR DESCRIPTION
**Pull request by ManWithBear**
Average channel open time (duration from tap to idle state) is 1.3s.
A lot of time takes data base updates and table reloads.
But in most cases such reloads are redundant because object haven;t been changed, only touched and marked as dirty.

By reseting dirty, but not changed, objects I was able to reduce open time from 1.3s to 0.6s.

Rest of time is initial layout setup (VC and message list).  It doesn't make much sense to invest time in this, since layout policy is about to change.

### CODE REVIEW

File: Sources/StreamChat/Database/DatabaseContainer.swift [hunk 1]
```diff
@@ -152,6 +152,14 @@ class DatabaseContainer: NSPersistentContainer {
             log.debug("Starting a database session.")
             do {
                 try actions(self.writableContext)
+                // If you touch ManagedObject and update one of it properties to same value
+                // Object will be marked as `updated` even it hasn't changed.
+                // By reseting such objects we remove updates that are not updates.
+                for object in self.writableContext.updatedObjects {
+                    if object.changedValues().isEmpty {
+                        self.writableContext.refresh(object, mergeChanges: false)
+                    }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DatabaseContainer.swift:161]: Do I understand it correctly, that `object.hasChanges` is set whenever you set a value on the object, but `object.changedValues()` actually evaluates the value and returns only when the value was actually changed?

Does it mean that we can actually revert the changes from https://github.com/GetStream/stream-chat-swift/pull/752 ? cc @b-onc

**Review by VojtaStavik — APPROVED** [review]

File: Sources/StreamChat/Database/DatabaseContainer.swift (hunk 1 above)
Comment (b-onc) [reply Sources/StreamChat/Database/DatabaseContainer.swift:161]: If that's ⬆️ the case, we definitely don't need that persisted hash workaround, this would be a much cleaner one.
Also, the existence of this PR proves that #752 did not resolve the issue fully.

Comment (ManWithBear) [reply Sources/StreamChat/Database/DatabaseContainer.swift:161]: Basically yes. `hasChanges` is just dirty flag, so whenever you touch object, it gets marked as with changes.
`changedValues` contains pairs of keys-values of actually changed values.
Here exist additional property `hasPersistentChangedValues` that should work as you would expect from `hasChanges`, but documentation is lucking, so I went with current solution

File: Sources/StreamChat/Database/DTOs/AttachmentDTO.swift [hunk 2]
```diff
@@ -41,6 +41,16 @@ class AttachmentDTO: NSManagedObject {
...
     @NSManaged var message: MessageDTO
     @NSManaged var channel: ChannelDTO
+
+    override func willSave() {
+        super.willSave()
+
+        // When attachment changed, we need to propagate this change up to holding message
+        if hasPersistentChangedValues, !message.hasChanges {
+            // this will not change object, but mark it as dirty, triggering updates
+            message.id = message.id
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:51]: How come this change is not removed by https://github.com/GetStream/stream-chat-swift/pull/802/commits/df7983b65dac26188897ea48caad9bf8cf4d0b03 ?

File: Sources/StreamChat/Database/DTOs/UserDTO.swift [hunk 3]
```diff
@@ -30,10 +33,24 @@ class UserDTO: NSManagedObject {
...
+        if hasPersistentChangedValues {
+            if !(currentUser?.hasChanges ?? false) {
+                // this will not change object, but mark it as dirty, triggering updates
+                currentUser?.unreadChannelsCount = currentUser?.unreadChannelsCount ?? 0
+            }
+            for member in members ?? [] {
+                guard !member.hasChanges else { continue }
+                // this will not change object, but mark it as dirty, triggering updates
+                let swiftBelieveMeIKnowWhatImDoing = member.channelRoleRaw
+                member.channelRoleRaw = swiftBelieveMeIKnowWhatImDoing
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO.swift:50]: I understand this is a workaround for the compiler warning. I wouldn't mind if you named it the way such that it's more obvious.

File: Sources/StreamChat/Database/DTOs/AttachmentDTO.swift [hunk 4]
```diff
@@ -41,6 +41,16 @@ class AttachmentDTO: NSManagedObject {
     
     @NSManaged var message: MessageDTO
     @NSManaged var channel: ChannelDTO
+
+    override func willSave() {
+        super.willSave()
+
+        // When attachment changed, we need to propagate this change up to holding message
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:48]: Can you add a test for this, please?

File: Sources/StreamChat/Database/DTOs/UserDTO.swift [hunk 5]
```diff
@@ -30,10 +33,24 @@ class UserDTO: NSManagedObject {
...
     override func willSave() {
         super.willSave()
-        
+
+        // When user changed, we need to propagate this change to members and current user
+        if hasPersistentChangedValues {
+            if !(currentUser?.hasChanges ?? false) {
+                // this will not change object, but mark it as dirty, triggering updates
+                currentUser?.unreadChannelsCount = currentUser?.unreadChannelsCount ?? 0
+            }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO.swift:45]: wouldn't this be more readable and simple using `if let currentUser = currentUser, currentUSer.hasChanges { }` ?

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 6]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
             NSSortDescriptor(key: "userRoleRaw", ascending: true)
         )
     }
+
+    func test_DTO_triggerMembersUpdate() throws {
+        // Arrange: Store member and user in database
+        let userId = UUID().uuidString
+        let channelId = ChannelId(type: .init(rawValue: "messaging"), id: UUID().uuidString)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:380]: nit: you can do:
```
let userId: UserId = .unique
let channelId: ChannelId = .unique
```

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 7]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
...
+
+        let payload: MemberPayload<NoExtraData> = .init(
+            user: userPayload,
+            role: .moderator,
+            createdAt: .init(timeIntervalSince1970: 4000),
+            updatedAt: .init(timeIntervalSince1970: 5000)
+        )
+
+        try database.writeSynchronously { session in
+            try! session.saveMember(payload: payload, channelId: channelId)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:392]: why `try!` here when the whole method is `throws`?

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 8]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
...
+        }
+
+        // Arrange: Observe changes on members
+        let observer = ListDatabaseObserver<MemberDTO, MemberDTO>(
+            context: database.viewContext,
+            fetchRequest: MemberDTO.member(userId, in: channelId),
+            itemCreator: { $0 }
+        )
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:400]: nit: You can use `EntityDatabaseObserver`, the functionality is similar it just makes your intentions clearer.

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 9]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
...
+        try observer.startObserving()
+
+        var receivedChanges: [ListChange<MemberDTO>]?
+        observer.onChange = { receivedChanges = $0 }
+
+        // Act: Update user
+        try database.writeSynchronously { session in
+            let loadedUser: UserDTO = try XCTUnwrap(session.user(id: userId))
+            loadedUser.name = "Jo"
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:409]: nit: did you mean Jo Jo? https://starwars.fandom.com/wiki/Jo_Jo

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 10]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
             NSSortDescriptor(key: "userRoleRaw", ascending: true)
         )
     }
+
+    func test_DTO_triggerMembersUpdate() throws {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:377]: The name of the method should reflect what you test. Maybe something like `test_changingUserDTO_triggersMemberDTOUpdate`?

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 11]
```diff
@@ -373,4 +373,82 @@ class UserDTO_Tests: XCTestCase {
...
+        try database.writeSynchronously { session in
+            let loadedUser: UserDTO = try XCTUnwrap(session.user(id: userId))
+            loadedUser.name = "Jo"
+        }
+
+        // Assert: Members should be updated
+        XCTAssertFalse(receivedChanges?.isEmpty ?? true)
+    }
+
+    func test_DTO_triggerCurrentUserUpdate() throws {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:416]: -> `test_changingUserDTO_triggersCurrentUserDTOUpdate` ?

**Review by VojtaStavik — COMMENTED** [review]
Looking good, I have mostly nit comments. One thing I don't understand is
> How come this change is not removed by df7983b ?

File: Sources/StreamChat/Database/DTOs/UserDTO.swift (hunk 5 above)
Comment (VojtaStavik) [reply Sources/StreamChat/Database/DTOs/UserDTO.swift:45]: WAIT, we shouldn't check `currentUser`'s changes here, right?

File: Sources/StreamChat/Database/DTOs/AttachmentDTO.swift (hunk 2 above)
Comment (ManWithBear) [reply Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:51]: `willSave` is called inside `context.save()`, so our "cleanup" logic is already performed.

File: Sources/StreamChat/Database/DTOs/AttachmentDTO.swift (hunk 4 above)
Comment (ManWithBear) [reply Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:48]: This behavior is already tested by `MessageSender`, e.g. in `test_sender_sendsMessage_withBothAttachmentSeedsAndUploadedAttachments`

Comment (VojtaStavik) [reply Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:48]: I see. Seems to me that having an extra test here for this specific behavior will be beneficial for future use cases. If you don't mind adding it it would great.

File: Sources/StreamChat/Database/DTOs/AttachmentDTO.swift (hunk 2 above)
Comment (VojtaStavik) [reply Sources/StreamChat/Database/DTOs/AttachmentDTO.swift:51]: Gotcha 👍

**Comment by VojtaStavik** [comment]
@ManWithBear All right, let's merge this!
